core/predicate_check.go

```diff
@@ -18,50 +18,90 @@ import (
 
 var ErrMissingPredicateContext = errors.New("missing predicate context")
 
-// CheckPredicates verifies the predicates of [tx] and returns the result. Returning an error invalidates the block.
-func CheckPredicates(rules params.Rules, predicateContext *precompileconfig.PredicateContext, tx *types.Transaction) (predicate.PrecompileResults, error) {
-	// Check that the transaction can cover its IntrinsicGas (including the gas required by the predicate) before
-	// verifying the predicate.
-	intrinsicGas, err := IntrinsicGas(tx.Data(), tx.AccessList(), tx.To() == nil, rules)
+// CheckBlockPredicates verifies the predicates of a block of transactions and
+// returns the results.
+//
+// Returning an error invalidates the block.
+func CheckBlockPredicates(
+	rules params.Rules,
+	predicateContext *precompileconfig.PredicateContext,
+	txs []*types.Transaction,
+) (predicate.BlockResults, error) {
+	var results predicate.BlockResults
```

Comment:
Also, we could have pre-allocated this map to the number of txs. Which would ensure we never need to resize the block results... But because most txs probably don't have predicates, I felt like that might not actually be a good optimization.